py/src/call.py

```diff
@@ -0,0 +1,427 @@
+        finally:
+            connection.rollback()
+
+            completed_at = time.time()
+            timings = {}
+
+            if parsed_at is not None and started_at < parsed_at:
+                timings["parsing"] = parsed_at - started_at
+
+            if parsed_at is not None and parsed_at < completed_at:
+                timings["execution"] = completed_at - parsed_at
+
+            out["timings"] = timings
+
+            print(json.dumps(out), file=output_file, flush=True)
+
+
+def parse_command(command, required, optional):
+    # it would be nice to use marshmallow but before we can lock with
+    # cairo-lang we cannot really add common dependencies
+    missing = required.keys() - command.keys()
+    assert len(missing) == 0, f"missing keys from command: {missing}"
+
+    extra = set()
+
+    converted = dict()
+
+    for k, v in command.items():
+        conv = required.get(k, None)
+        if conv is None:
+            conv = optional.get(k, None)
+        if conv is None:
+            extra += k
+            continue
+
+        try:
+            converted[k] = conv(v)
+        except Exception:
+            raise InvalidInput(k)
+
+    assert len(extra) == 0, f"extra keys from command: {extra}"
+
+    return converted
+
+
+def int_hash_or_latest(s):
+    if type(s) == int:
+        return s
+    if s == "latest":
+        return s
+    if s == "pending":
+        # this is allowed in the rpc api but pathfinder doesn't create the blocks
+        # this should had never come to us
+        raise NoSuchBlock(s)
+    assert s[0:2] == "0x"
+    return len_safe_hex(s)
+
+
+def hash_or_int(s):
+    if type(s) == int:
+        return s
+    if s.startswith("0x"):
+        return int.from_bytes(len_safe_hex(s), "big")
+    return int(s, 10)
+
+
+def len_safe_hex(s):
+    """
+    Over at the RPC side which pathfinder supports, sometimes hex without
+    leading zeros is needed, so they could come over to python as well.
+    bytes.fromhex doesn't support odd length hex, it gives a non-hex character
+    error.
+    """
+    if s.startswith("0x"):
+        s = s[2:]
+    if len(s) % 2 == 1:
+        s = "0" + s
+    return bytes.fromhex(s)
+
+
+def string_or_int(s):
+    if type(s) == int:
+        return s
+
+    if type(s) == str:
+        if s.startswith("0x"):
+            return hash_or_int(s)
+        # not sure if this should be supported but strings get ran through the
+        # truncated keccak
+        return s
+
+    raise TypeError(f"expected string or int, not {type(s)}")
+
+
+def list_of_hash_or_int(s):
+    assert type(s) == list, f"Expected list, got {type(s)}"
+    return list(map(hash_or_int, s))
+
+
+def check_schema(connection):
+    global first
+    assert connection.in_transaction
+    cursor = connection.execute("select user_version from pragma_user_version")
+    assert cursor is not None, "there has to be an user_version defined in the database"
+
+    [version] = next(cursor)
+    return version == 2
```

Comment:
So we need to remember to bump this value here when fiddling with the schema on the rust side, maybe it's worth pointing somewhere around the migration fn in Rust  :shrug:

Reply:
Aah yes, this is true. This happens if you don't:

```
🏁 Starting node.
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Launch first python executor

Caused by:
    0: Failed to start python subprocess
    1: failed to read ready from python process', crates/pathfinder/src/bin/pathfinder.rs:26:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

So it happens on the first launch, and it will not try to re-launch this every 1s, which means this is sort of "self-tested". It can start happening between executed commands, it could be more lenient (based on the implicit schema assumptions) but I just think it's more self-testable if we are strict as possible. We should only ever have one revision of the schema, and this can easily be made into `return version in [2, 3, 4, 5, 6]`.

Reply:
This should actually be a test case. Yay for finding more!